multicluster/controllers/multicluster/node_controller.go

```diff
@@ -73,67 +78,215 @@ func NewNodeReconciler(
 
 func (r *NodeReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
 	klog.V(2).InfoS("Reconciling Node", "node", req.Name)
+	if !r.initialized {
+		if err := r.initializeActiveGateway(); err != nil {
+			return ctrl.Result{}, err
+		}
+		r.initialized = true
+	}
 	gw := &mcsv1alpha1.Gateway{
 		ObjectMeta: metav1.ObjectMeta{
 			Name:      req.Name,
 			Namespace: r.namespace,
 		},
 	}
-	// When the Node is annotated with 'multicluster.antrea.io/gateway=true' as a Gateway:
-	//   - Delete the Gateway if the Node is deleted
-	//   - Update the Gateway if Node's InternalIP or GatewayIP is updated
-	//   - Create a new Gateway if there is no existing Gateway
+
 	node := &corev1.Node{}
 	if err := r.Client.Get(ctx, req.NamespacedName, node); err != nil {
 		if apierrors.IsNotFound(err) {
-			err := r.Client.Delete(ctx, gw, &client.DeleteOptions{})
-			return ctrl.Result{}, client.IgnoreNotFound(err)
+			if r.activeGateway == req.Name || r.activeGateway == "" {
+				if err := r.recreateActiveGateway(ctx, gw); err != nil {
+					return ctrl.Result{}, err
+				}
+				return ctrl.Result{}, nil
+			}
 		}
 		klog.ErrorS(err, "Failed to get Node", "node", req.Name)
-		return ctrl.Result{}, err
+		return ctrl.Result{}, client.IgnoreNotFound(err)
 	}
 
-	_, isGW := node.Annotations[common.GatewayAnnotation]
-	var err error
-	gwNamespacedName := types.NamespacedName{
-		Name:      node.Name,
-		Namespace: r.namespace,
-	}
+	_, hasGWAnnotation := node.Annotations[common.GatewayAnnotation]
+	nodeIsReady := isReadyNode(node)
+	noActiveGateway := r.activeGateway == ""
+	isActiveGateway := r.activeGateway == node.Name
 
-	// TODO: cache might be stale. Need to revisit here and other reconcilers to
-	// check if we can improve this with 'Owns' or other methods.
-	var gwIP, internalIP string
-	if isGW {
-		if internalIP, gwIP, err = r.getGatawayNodeIP(node); err != nil {
-			klog.ErrorS(err, "There is no valid Gateway IP for Node, will retry later when there is any new Node update", "node", node.Name)
+	if !nodeIsReady {
+		if isActiveGateway {
+			if err := r.recreateActiveGateway(ctx, gw); err != nil {
+				return ctrl.Result{}, err
+			}
 			return ctrl.Result{}, nil
 		}
+		if hasGWAnnotation {
```

Comment:
I feel much easier to understand if we move this block out of `if !nodeIsReady`.

Reply:
Done